### PR TITLE
Use directories for autoloading

### PR DESCRIPTION
**Pull request by jcgruenhage**
fixes https://github.com/nushell/nushell/issues/13378

# Description

This PR tries to improve usage of system APIs to determine the location of vendored autoload files.

# User-Facing Changes

The paths listed in #13180 and #13217 are changing. This has not been part of a release yet, so arguably the user facing changes are only to unreleased features anyway.

# Tests + Formatting

Haven't done, but if someone wants to help me here, I'm open to doing it. I just don't know how to properly test this.

# After Submitting

### CODE REVIEW

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [inline crates/nu-protocol/src/eval_const.rs:283]: I'd like to receive some feedback on this. What do y'all think? https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW1 is the relevant documenation ftr

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [inline crates/nu-protocol/src/eval_const.rs:285]: Also, what do people think about this?

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [inline crates/nu-protocol/src/eval_const.rs:296]: Same here?

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [inline crates/nu-protocol/src/eval_const.rs:286]: I doubt this will be allowed. It's preferred to get env var from stack or engine_state.

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [inline crates/nu-protocol/src/eval_const.rs:297]: I'm not a fan of using winapi. We should use the window-rs crate if we want to do things this way.

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [inline crates/nu-protocol/src/eval_const.rs:203]: I'm not sure I like this because we don't know what order things get sourced/used.

**Comment by fdncred** [comment]
I've made a few comments. In general, I'm not sure what makes this way better than the previous way, especially since you've now removed the ability to make it configurable at compile time.

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:203]: We don't? Aren't lists ordered? Maybe this is just something to extend documentation on?

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:297]: nushell already uses `dirs_next` and `dirs_sys_next`, which under the hood use winapi. Introducing `windows-rs` for just this single call would be blowing up the dependency tree quite a bit without much reason.

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:286]: Well, that's why I asked that question in a todo comment right above. There's certain benefits to using `std::env::var_os`, such as also working with paths that are not valid unicode, and anything using `dirs_next` will use not us the stack or engine state anyway. I *can* switch to the engine_state if that's preferrable, but I'm just not sure on the trade-off here, and the inconsistency it'll introduce.

**Comment by jcgruenhage** [comment]
> especially since you've now removed the ability to make it configurable at compile time

I did intent to put this in, but forgot about it and haven't revisited the whole diff yet, which is why this is marked as a draft still.

I'll be adding it later, and mark it as ready for review once I'm satisfied with the problems that I'm already aware of.

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [reply crates/nu-protocol/src/eval_const.rs:286]: As I recall, we try not to call into the OS environment at all in order to try and keep our thread-safety intact. Maybe other core-team members have opinions here.

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [reply crates/nu-protocol/src/eval_const.rs:297]: Maybe we should use something more modern that dirs-next?

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [reply crates/nu-protocol/src/eval_const.rs:203]: Not that I know of. Doesn't that imply some internal sorting somewhere in nushell? If so, I'm pretty sure that doesn't happen.

Comment (LordMZTE) [reply crates/nu-protocol/src/eval_const.rs:203]: Why would this be unordered? `Value::List` just wraps an (ordered) `Vec`. The return value from `get_vendor_autoload_dir` is also sorted the same as the values in the original environment variable, as expected.

File: crates/nu-protocol/src/eval_const.rs
Comment (LordMZTE) [inline crates/nu-protocol/src/eval_const.rs:266]: This is a little nitpick, but perhaps we should rename this to `get_vendor_autoload_dirs` as it now returns multiple values.

**Comment by jcgruenhage** [comment]
speaking of `dirs_next` earlier in this thread: Any reason that's still used? That's a three year old fork from then the `dirs` maintainer had put the project on hold for a short while, but `dirs` is back and the more commonly used variant by far.

**Review by LordMZTE — CHANGES_REQUESTED** [review]

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:297]: I'm fine with also switching over `dirs_next` to `dirs`, which funnily enough is the more maintained version of these two, and then using the `windows-sys` crate instead of `winnapi`. But that's a bigger change.

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [reply crates/nu-protocol/src/eval_const.rs:203]: I'm not sure I understand the problem here either - lists are ordered, we should parse `XDG_DATA_DIRS` in order.

Should it be expected that things on the left side of the list will load first or last? If the default is `/usr/local/share:/usr/share`, that kind of implies that the beginning of the list should load last (you'd want `/usr/local` stuff to be able to replace `/usr` stuff.)

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [reply crates/nu-protocol/src/eval_const.rs:283]: This is where it gets a bit messy. We decided to go with `~/Library/Application Support` for our config, so it would kind of make sense to go with `/Library/Application Support` for the system level config too.

However, most users on macOS of TUI apps expect them to basically use XDG dirs like `~/.config` and I've seen a lot of people set `XDG_CONFIG_HOME` after we started supporting that to avoid `~/Library`. I think the perception is that `~/Library` is the right place for macOS graphical applications to put things, but that the Terminal should behave like any other Unix-like environment

I think just for consistency for now we might as well go with `/Library/Application Support` and *also* support `XDG_DATA_DIRS`. I've also heard from Windows folks that they would like to be able to configure stuff as well and they don't mind using the XDG spec environment variables to do that, but that kind of creates a problem for `XDG_DATA_DIRS` because it's colon separated, we'd have to be non-standard and allow semicolon to be used like Windows `Path`.

And I think compile time configuration is important to allow Homebrew or whatever else to do what they want. I'm not sure if they rely on setting `XDG_DATA_DIRS` but I do know they set `PREFIX` at compile time.

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [reply crates/nu-protocol/src/eval_const.rs:286]: We just generally don't support non-Unicode paths. It's a known issue with how we do things, but it's probably not going to be resolved in this PR.

I think it is okay to use the true environment variable at load time, though, as we never set these and it is not possible to change the vendor autoload dir from Nushell configuration (that just wouldn't make sense).

We do need to load env variables from `Stack` at runtime but this is a bit of a different situation

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [inline crates/nu-protocol/src/eval_const.rs:302]: Does it make sense to have a user-local vendor autoload dir? Hmm... maybe?

I was thinking it would make more sense to rework our user config to also be based on loading files from a directory instead of doing that

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:283]: At least for the autoload dirs, we don't have to decide on either but can instead just go for both without any trouble? The more, the merrier. We just have to decide on a precedence.

> And I think compile time configuration is important to allow Homebrew or whatever else to do what they want. I'm not sure if they rely on setting XDG_DATA_DIRS but I do know they set PREFIX at compile time.

So, what do you suggest here? https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables says 

> If $XDG_DATA_DIRS is either not set or empty, a value equal to /usr/local/share/:/usr/share/ should be used.

Should we just replace `/usr` with `PREFIX` for both of those? Add an additional one with `PREFIX`, and dedupe the list in the end? I think that would be my preferred option.

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:203]: Yep, it should load right to left, so that the precedence is left to right. I still need to reverse the loading order for that to actually happen.

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs:302]: At least it was suggested in https://github.com/nushell/nushell/issues/13378 to also load completions/extensions from `~/.local/share/nushell/vendor/autoload`. Reading both `XDG_DATA_DIRS` (system dirs) and `XDG_DATA_HOME` (user dir) is accordance with the XDG spec here and seems like a reasonable choice.

https://github.com/nushell/nushell/issues/13378#issuecomment-2228304411 even explicitly invited PRs on this.

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [reply crates/nu-protocol/src/eval_const.rs:283]: I suppose we wouldn't want to do `/usr/local/local` if `PREFIX = /usr/local` though, so it might be reasonable to just special case that

But I think that does sound fine to me

File: crates/nu-protocol/src/eval_const.rs
Comment (devyn) [reply crates/nu-protocol/src/eval_const.rs:302]: Okay, I don't really see it being a problem, though before we rework user config it might be weird if people start using that instead just because they prefer how it's laid out

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs]: I've opened a PR for replacing dirs_next with dirs in https://github.com/nushell/nushell/pull/13384, rebased this on top of it and switched to using windows-rs.

If someone could test this, I'd appreciate it, but considering this is done, I'll be marking this thread as resolved either way.

File: crates/nu-protocol/src/eval_const.rs
Comment (jcgruenhage) [reply crates/nu-protocol/src/eval_const.rs]: done

**Comment by jcgruenhage** [comment]
I think I have resolved most of the open points by now. The only "problem" I'm still aware of is env var handling, but the feedback on that topic was mixed, so I've not switched over to getting env vars from engine_state yet.

File: crates/nu-protocol/src/eval_const.rs
Comment (IanManske) [reply crates/nu-protocol/src/eval_const.rs]: Yeah, `setup_config`, `get_vendor_autoload_dir`, and `read_vendor_autoload_files` should all only be called at shell startup (before any user code is run), so it's fine to use the process's environment variables instead of using the engine state and stack.

File: crates/nu-protocol/src/eval_const.rs
Comment (IanManske) [inline crates/nu-protocol/src/eval_const.rs:310]: I think it's possible to avoid using the `unsafe` method:
```suggestion
    #[cfg(unix)]
    {
        use std::os::unix::ffi::OsStrExt;
        
        std::env::var_os("XDG_DATA_DIRS")
```
(continued below)

File: crates/nu-protocol/src/eval_const.rs
Comment (IanManske) [inline crates/nu-protocol/src/eval_const.rs:324]:
```suggestion
                let os_str = std::ffi::OsStr::from_bytes(split);
```

**Comment by jcgruenhage** [comment]
Thanks @IanManske, feedback has been included.

I've also added deduplication, so that directories aren't added multiple times if they happen to already be on the list. Short verification using `cargo run -- --log-level warn` looks good to me:

```
2024-07-16 07:10:10.063 AM [WARN ] nu::config_files: read_vendor_autoload_files() src/config_files.rs:199:9
2024-07-16 07:10:10.063 AM [WARN ] nu::config_files: read_vendor_autoload_files: /usr/share/nushell/vendor/autoload
2024-07-16 07:10:10.063 AM [WARN ] nu::config_files: AutoLoading: "/usr/share/nushell/vendor/autoload/minijinja-cli.nu"
2024-07-16 07:10:10.068 AM [WARN ] nu::config_files: read_vendor_autoload_files: /usr/local/share/nushell/vendor/autoload
2024-07-16 07:10:10.068 AM [WARN ] nu::config_files: read_vendor_autoload_files: /var/lib/flatpak/exports/share/nushell/vendor/autoload
2024-07-16 07:10:10.068 AM [WARN ] nu::config_files: read_vendor_autoload_files: /home/jcgruenhage/.local/share/flatpak/exports/share/nushell/vendor/autoload
2024-07-16 07:10:10.068 AM [WARN ] nu::config_files: read_vendor_autoload_files: /home/jcgruenhage/.local/share/nushell/vendor/autoload
```

**Comment by jcgruenhage** [comment]
Important notice: I've only tested this on Linux. It'd probably be good if someone tried this on Windows and macOS, I've implemented these blindly so far.

**Comment by jcgruenhage** [comment]
Got a friend to run it on macOS, and their local nix install is nicely discovered via `XDG_DATA_DIRS`:

```
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files() src/config_files.rs:199:9
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /nix/var/nix/profiles/default/share/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /run/current-system/sw/share/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /etc/profiles/per-user/redacted/share/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /Users/redacted/.nix-profile/share/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /Library/Application Support/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /Users/redacted/.local/share/nushell/vendor/autoload
2024-07-16 11:22:40.560 AM [WARN ] nu::config_files: read_vendor_autoload_files: /Users/redacted/Library/Application Support/nushell/vendor/autoload
```

File: crates/nu-protocol/src/eval_const.rs
Comment (fdncred) [inline crates/nu-protocol/src/eval_const.rs:261]: `%PROGRAM_DATA%` isn't a thing but there is `%ProgramData%`

**Comment by fdncred** [comment]
I tried testing on windows but it won't compile. It's missing the dirs-sys and windows crates. Ping me when this is fixed and I can try again.

**Comment by jcgruenhage** [comment]
Fixing this without access to a windows machine is a bit difficult, but I'll look into getting some Windows 11 machine going.

**Comment by fdncred** [comment]
@jcgruenhage I added the crates (but didn't commit that change) but there were still errors. I just don't have time to mess with it right now. Maybe this will help.
```
❯ cargo r -- --log-level warn                                                                                                                                                                        08:07:21 AM
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
   Compiling nu-protocol v0.95.1 (C:\Users\username\source\repos\forks\nushell\crates\nu-protocol)
   Compiling nu v0.95.1 (C:\Users\username\source\repos\forks\nushell)
error[E0433]: failed to resolve: use of undeclared crate or module `windows`
   --> crates\nu-protocol\src\eval_const.rs:337:28
    |
337 |     dirs_sys::known_folder(windows::Win32::UI::Shell::FOLDERID_ProgramData)
    |                            ^^^^^^^ use of undeclared crate or module `windows`

error[E0631]: type mismatch in closure arguments
   --> crates\nu-protocol\src\eval_const.rs:339:14
    |
263 |     let into_autoload_path_fn = |mut path: PathBuf| {
    |                                 ------------------- found signature defined here
...
339 |         .map(into_autoload_path_fn)
    |          --- ^^^^^^^^^^^^^^^^^^^^^ expected due to this
    |          |
    |          required by a bound introduced by this call
    |
    = note: expected closure signature `fn(&std::path::PathBuf) -> _`
               found closure signature `fn(std::path::PathBuf) -> _`
note: required by a bound in `map`
   --> C:\Users\username\.rustup\toolchains\1.77.2-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\iter\traits\iterator.rs:805:33
    |
802 |     fn map<B, F>(self, f: F) -> Map<Self, F>
    |        --- required by a bound in this associated function
...
805 |         F: FnMut(Self::Item) -> B,
    |                                 ^ required by this bound in `Iterator::map`
help: consider wrapping the function in a closure
    |
339 |         .map(|arg0: &std::path::PathBuf| into_autoload_path_fn(*arg0))
    |              +++++++++++++++++++++++++++                      +++++++
help: consider adjusting the signature so it borrows its argument
    |
263 |     let into_autoload_path_fn = |mut path: &PathBuf| {
    |                                            +

error[E0599]: the method `for_each` exists for struct `Map<Iter<'_, PathBuf>, {closure@eval_const.rs:263:33}>`, but its trait bounds were not satisfied
   --> crates\nu-protocol\src\eval_const.rs:340:10
    |
263 |       let into_autoload_path_fn = |mut path: PathBuf| {
    |                                   ------------------- doesn't satisfy `<_ as FnOnce<(&PathBuf,)>>::Output = _`
...
337 | /     dirs_sys::known_folder(windows::Win32::UI::Shell::FOLDERID_ProgramData)
338 | |         .iter()
339 | |         .map(into_autoload_path_fn)
340 | |         .for_each(&mut append_fn);
    | |         -^^^^^^^^ method cannot be called due to unsatisfied trait bounds
    | |_________|
    |
    |
   ::: C:\Users\username\.rustup\toolchains\1.77.2-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\iter\adapters\map.rs:62:1
    |
62  |   pub struct Map<I, F> {
    |   -------------------- doesn't satisfy `_: Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `<{closure@crates\nu-protocol\src\eval_const.rs:263:33: 263:52} as FnOnce<(&std::path::PathBuf,)>>::Output = _`
            which is required by `std::iter::Map<std::option::Iter<'_, std::path::PathBuf>, {closure@crates\nu-protocol\src\eval_const.rs:263:33: 263:52}>: Iterator`
            `std::iter::Map<std::option::Iter<'_, std::path::PathBuf>, {closure@crates\nu-protocol\src\eval_const.rs:263:33: 263:52}>: Iterator`
            which is required by `&mut std::iter::Map<std::option::Iter<'_, std::path::PathBuf>, {closure@crates\nu-protocol\src\eval_const.rs:263:33: 263:52}>: Iterator`

Some errors have detailed explanations: E0433, E0599, E0631.
For more information about an error, try `rustc --explain E0433`.
error: could not compile `nu-protocol` (lib) due to 3 previous errors
```

**Comment by jcgruenhage** [comment]
I've got a Windows 11 machine running now, but I'm afraid I won't finish this today. In #13384 it sounded like you'd want things to get in quickly, or *after* the next release, does this also apply to this PR?

**Comment by fdncred** [comment]
> I've got a Windows 11 machine running now, but I'm afraid I won't finish this today. In #13384 it sounded like you'd want things to get in quickly, or _after_ the next release, does this also apply to this PR?

yes, but we can wait until after the release next tuesday if that's better for you. we just don't want to make fast changes and then have many things broken for the next release and have to do patches.

**Comment by jcgruenhage** [comment]
I've gotten it to compile, I want to get this over, I have more time on my hands this week then next week and after that. I'll try to commit my fixes later.

**Comment by kubouch** [comment]
Just to comment on the environment variables. The current setup (reading from `std::env` works if the intent is to not be able to change the `NU_VENDOR_AUTOLOAD_DIR` at runtime and only use whatever was inherited from the parent process. Since the autoload dir is read [after the config](https://github.com/nushell/nushell/blob/b66671d339325dc34f96af4d893f9f94cafc4f0f/src/config_files.rs#L281), doing `$env.NU_VENDOR_AUTOLOAD_DIR` inside any of the config files won't have any effect. Same for the XDG_xxx env vars. It's OK for me, though at some point, we could change it to reading the env var from the Stack, thus making it user-settable, and add a new `autoload` command that would reload the files.

**Comment by jcgruenhage** [comment]
@fdncred I have fixed windows, at least on my VM it now works as intended. From my perspective, this is *probably* good to go now.

@kubouch yep, this is something to consider again later when `autoload` as a command comes in. Good point to explicitly mention it though :)

**Comment by jcgruenhage** [comment]
I've deliberately ordered them in reverse, so that first in the list means higher precedence. That feels more intuitive to me. That higher precedence is achieved by being loaded later is more of a side effect in my opinion.

If people feels strongly about this, we can also reorder it.

**Comment by jcgruenhage** [comment]
Other variables (like XDG_DATA_DIRS and PATH) are also in order of precedence, instead of order of loading, in case that matters for this decision?

**Comment by fdncred** [comment]
ya, imo they should all be the same, listed in order that they're loaded.

**Comment by jcgruenhage** [comment]
But they are all in order of precedence out there? Why deviate from this established practice?

**Comment by jcgruenhage** [comment]
What "order of loading" even means is more of an implementation detail in my opinion. For something like PATH, you could go through the directories in order and look for a binary until you've found it. Then order of precedence and order of loading are the same. You could also build a lookup table, where binaries found later replace previously found binaries. Then order of precedence is reversed to what order of loading is. Going for semantic meaning as opposed to reflecting how internals are implemented is preferrable here I'd say.

**Comment by fdncred** [comment]
All I'm saying is that we are storing the paths in a certain order like this
image
I think that should mean that directory 0's contents are loaded first, then directory 1's contents are loaded.

**Comment by devyn** [comment]
@fdncred I think it's kind of the opposite situation from `$env.PATH`, in that case you want to search the first ones first because you want those to take precedence, but in the case of load order, whatever you load last has the most precedence. You'd want user config to be loaded first. Because `PATH` is usually prepended to, it would make sense to also follow that pattern of prepending to that list if you're going to add something new.

**Comment by fdncred** [comment]
I think it's wrong and unintuitive but I'm fine doing things different that I prefer them as long as someone documents that directory load order is the reverse as listed.

**Comment by devyn** [comment]
Of course the alternative would be to just treat `XDG_DATA_DIRS` in the logical order (i.e. reverse it) to match the expected order of precedence, but present  `$nu.vendor-autoload-dir` as being in load order (first loaded first)

**Comment by jcgruenhage** [comment]
> I think it's wrong and unintuitive but I'm fine doing things different that I prefer them as long as someone documents that directory load order is the reverse as listed.

@fdncred: Are you also fine with it as long someone documents that the list is sorted by precedence, so items from the directories listed first take precedence over those in directories listed later?

That is what I would document. If I think about loading the files, sure, it's unintuitive, but I thinking about precedence is more intuitive in my opinion, as load order is just an implementation detail IMO.

If this gets merged, I'll make sure to open a PR with appropriate documentation in https://github.com/nushell/nushell.github.io.

**Comment by NotTheDr01ds** [comment]
If (big if) I'm reading things correctly here, I agree (pretty strongly, I *think*) with @fdncred that the order should be load order.

If `print` statements appear in multiple files, then the order that they are displayed to the user during startup will be according to *load order*, correct? If so, then that's the order that Nushell should list them.

I think that most users naturally understand (or will, with experience) that changing a variable (e.g. `PATH`) in a later-loaded file will  take precedence over earlier changes.

**Comment by fdncred** [comment]
> Are you also fine with it as long someone documents that the list is sorted by precedence

Not really. I'm not trying just to be contrary but imo talking about precedence is confusing. I think a nushell user will understand something like, `"auto-load directories are processed in the order that they're listed in $nu.vendor-autoload-dir, meaning the directory at index 0 is processed first, then directory at index 1, and so on."`. Once you start saying something takes precedence over the other, it just confuses the issue because you have to think more about it.

If core-team members disagree with me, they can land this PR. I won't have a problem with it. I understand both ways. I'm just thinking more about our users and like to try and keep-it-simple as much as possible.

**Comment by jcgruenhage** [comment]
Leaking implementation details such as processing order and telling people what implications that has is more confusing and harder to think about than "if something is defined in multiple of these folders, the definition from the folder higher in this list will win".

Either way, this is not a feature that people should need to think about at all. Autoloading and adding some configurability for the folders is so that distributions can properly ship completions and such automatically, so that users *don't have to think and configure stuff manually*.

In the end though: I don't care enough to continue arguing about this, and I think all the points have already been made anyway. Having someone from the core team making a decision here would be nice. If the decision is that we should reverse the order exposed to the user, then I will update this PR accordingly.

**Comment by fdncred** [comment]
I think it's important to know the order, but it is also important to understand that if in more-than-one of your folders you have "some_custom_command" defined, the last one sourced wins since it shadows any previous versions.

**Comment by devyn** [comment]
Hi @jcgruenhage,

We've discussed this as part of our core team meeting today and arrived at a consensus that:

- `$nu.vendor-autoload-dir` should be in load order
- `XDG_DATA_DIRS` should be reversed to insert into this, so that the expected precedence is preserved.

**Comment by jcgruenhage** [comment]
@devyn The PR has been adjusted accordingly.

**Comment by devyn** [comment]
Thanks! And thanks for testing this on Windows too. LGTM.

**Pull request merged**
